{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n"
     ]
    }
   ],
   "source": [
    "import os as os\n",
    "import sys as sys\n",
    "import re as re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json as json\n",
    "import pickle\n",
    "import pathlib\n",
    "from Bio import SeqIO\n",
    "\n",
    "pd.options.display.max_columns = 100\n",
    "pd.options.display.min_rows = None\n",
    "pd.options.display.max_rows = 20\n",
    "pd.options.display.max_colwidth = 100\n",
    "\n",
    "from config import MATERIALS_PATH, RESULTS_PATH\n",
    "\n",
    "FEATURE_PATH = MATERIALS_PATH.joinpath('mart_export.txt')\n",
    "\n",
    "UNSPLICED_PATH = MATERIALS_PATH.joinpath('martquery_0219183013_823.txt')\n",
    "UNSPLICED_RESULT_PATH = MATERIALS_PATH.joinpath('gene_symbol_dna_sequence_unspliced.pkl')\n",
    "\n",
    "EXON_PATH = MATERIALS_PATH.joinpath('martquery_0225145730_334.txt')\n",
    "EXON_RESULT_PATH = MATERIALS_PATH.joinpath('gene_symbol_dna_sequence_exon.pkl')\n",
    "\n",
    "PROTEIN_FASTA_PATH = MATERIALS_PATH.joinpath('uniprot-compressed_true_download_true_format_fasta_includeIsoform_tr-2023.02.17-22.00.02.49.fasta')\n",
    "PROTEIN_RESULT_PATH = MATERIALS_PATH.joinpath('gene_symbol_protein_sequences.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000064336</td>\n",
       "      <td>mt-Tf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000064337</td>\n",
       "      <td>mt-Rnr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000064338</td>\n",
       "      <td>mt-Tv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000064339</td>\n",
       "      <td>mt-Rnr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSMUSG00000064340</td>\n",
       "      <td>mt-Tl1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSMUSG00000064341</td>\n",
       "      <td>mt-Nd1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSMUSG00000064342</td>\n",
       "      <td>mt-Ti</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENSMUSG00000064343</td>\n",
       "      <td>mt-Tq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENSMUSG00000064344</td>\n",
       "      <td>mt-Tm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ENSMUSG00000064345</td>\n",
       "      <td>mt-Nd2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84715</th>\n",
       "      <td>ENSMUSG00000079005</td>\n",
       "      <td>Gm14147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84716</th>\n",
       "      <td>ENSMUSG00000085805</td>\n",
       "      <td>Gm14145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84717</th>\n",
       "      <td>ENSMUSG00000119215</td>\n",
       "      <td>Gm22517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84718</th>\n",
       "      <td>ENSMUSG00000067352</td>\n",
       "      <td>Gm14149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84719</th>\n",
       "      <td>ENSMUSG00000087337</td>\n",
       "      <td>Gm14144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84720</th>\n",
       "      <td>ENSMUSG00000089575</td>\n",
       "      <td>Gm23005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84721</th>\n",
       "      <td>ENSMUSG00000119806</td>\n",
       "      <td>Gm24077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84722</th>\n",
       "      <td>ENSMUSG00000087609</td>\n",
       "      <td>4930442J19Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84723</th>\n",
       "      <td>ENSMUSG00000119525</td>\n",
       "      <td>Gm23261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84724</th>\n",
       "      <td>ENSMUSG00000083391</td>\n",
       "      <td>Gm14148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57010 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene stable ID      Gene name\n",
       "0      ENSMUSG00000064336          mt-Tf\n",
       "3      ENSMUSG00000064337        mt-Rnr1\n",
       "7      ENSMUSG00000064338          mt-Tv\n",
       "10     ENSMUSG00000064339        mt-Rnr2\n",
       "14     ENSMUSG00000064340         mt-Tl1\n",
       "18     ENSMUSG00000064341         mt-Nd1\n",
       "22     ENSMUSG00000064342          mt-Ti\n",
       "25     ENSMUSG00000064343          mt-Tq\n",
       "28     ENSMUSG00000064344          mt-Tm\n",
       "31     ENSMUSG00000064345         mt-Nd2\n",
       "...                   ...            ...\n",
       "84715  ENSMUSG00000079005        Gm14147\n",
       "84716  ENSMUSG00000085805        Gm14145\n",
       "84717  ENSMUSG00000119215        Gm22517\n",
       "84718  ENSMUSG00000067352        Gm14149\n",
       "84719  ENSMUSG00000087337        Gm14144\n",
       "84720  ENSMUSG00000089575        Gm23005\n",
       "84721  ENSMUSG00000119806        Gm24077\n",
       "84722  ENSMUSG00000087609  4930442J19Rik\n",
       "84723  ENSMUSG00000119525        Gm23261\n",
       "84724  ENSMUSG00000083391        Gm14148\n",
       "\n",
       "[57010 rows x 2 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(FEATURE_PATH, sep='\\t')\n",
    "\n",
    "df = df[['Gene stable ID', 'Gene name']]\n",
    "\n",
    "df = df.drop_duplicates()\n",
    "\n",
    "_df_features = df.copy()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "for datum in SeqIO.parse(UNSPLICED_PATH, \"fasta\"):\n",
    "    record = {}\n",
    "    record['Gene stable ID'] = datum.description\n",
    "    record['Sequence'] = str(datum.seq)\n",
    "    records.append(record)\n",
    "\n",
    "df = pd.DataFrame(records)\n",
    "\n",
    "df = _df_features.merge(df, how='outer')\n",
    "\n",
    "assert not df['Gene stable ID'].duplicated().any()\n",
    "\n",
    "df.to_pickle(UNSPLICED_RESULT_PATH)\n",
    "\n",
    "_df_unspliced = df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000064336</td>\n",
       "      <td>mt-Tf</td>\n",
       "      <td>GTTAATGTAGCTTAATAACAAAGCAAAGCACTGAAAATGCTTAGATGGATAATTGTATCCCATAAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000064337</td>\n",
       "      <td>mt-Rnr1</td>\n",
       "      <td>AAAGGTTTGGTCCTGGCCTTATAATTAATTAGAGGTAAAATTACACATGCAAACCTCCATAGACCGGTGTAAAATCCCTTAAACATTTACTTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000064338</td>\n",
       "      <td>mt-Tv</td>\n",
       "      <td>CATAGTGTAGCTTAATATTAAAGCATCTGGCCTACACCCAGAAGATTTCATGACCAATGAACACTCTGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000064339</td>\n",
       "      <td>mt-Rnr2</td>\n",
       "      <td>ACTAATCCTAGCCCTAGCCCTACACAAATATAATTATACTATTATATAAATCAAAACATTTATCCTACTAAAAGTATTGGAGAAAGAAATTCGTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000064340</td>\n",
       "      <td>mt-Tl1</td>\n",
       "      <td>ATTAGGGTGGCAGAGCCAGGAAATTGCGTAAGACTTAAAACCTTGTTCCCAGAGGTTCAAATCCTCTCCCTAATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000064341</td>\n",
       "      <td>mt-Nd1</td>\n",
       "      <td>GTGTTCTTTATTAATATCCTAACACTCCTCGTCCCCATTCTAATCGCCATAGCCTTCCTAACATTAGTAGAACGCAAAATCTTAGGGTACATACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000064342</td>\n",
       "      <td>mt-Ti</td>\n",
       "      <td>AGAAATATGTCTGATAAAAGAATTACTTTGATAGAGTAAATTATAGAGGTTCAAGCCCTCTTATTTCTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000064343</td>\n",
       "      <td>mt-Tq</td>\n",
       "      <td>TAGGATAAGGTGTTTAGGTAGCACGGAGAATTTTGAATTCTTAAGTGTAGGTTCAATTCCTATTGTCCTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000064344</td>\n",
       "      <td>mt-Tm</td>\n",
       "      <td>AGTAAGGTCAGCTAATTAAGCTATCGGGCCCATACCCCGAAAACGTTGGTTTAAATCCTTCCCGTACTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000064345</td>\n",
       "      <td>mt-Nd2</td>\n",
       "      <td>ATAAATCCTATCACCCTTGCCATCATCTACTTCACAATCTTCTTAGGTCCTGTAATCACAATATCCAGCACCAACCTAATACTAATATGAGTAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57000</th>\n",
       "      <td>ENSMUSG00000079005</td>\n",
       "      <td>Gm14147</td>\n",
       "      <td>GGAGAGTGGTTTGGTGGTGGTTTGGGAGACGTGGACTGTGGTGATAATTTTGGATTTGGATACTGGTAACGTTTTGGACTCAGATTCTGGCAAAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57001</th>\n",
       "      <td>ENSMUSG00000085805</td>\n",
       "      <td>Gm14145</td>\n",
       "      <td>TGAGAACAGTTTCATCTGCATGGTGAACAGTGCAGGAGATGCCTGCCAAGCCCTAACACAGTAGACTGAAGATGGCCACACAGGTTATGCTTCCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57002</th>\n",
       "      <td>ENSMUSG00000119215</td>\n",
       "      <td>Gm22517</td>\n",
       "      <td>CCCACACCTAGAGGCATTGTTGCTTTGGCTCTGTGTTCCATGTGTATCAGTAGAAGGCAGAGAAGAGAGAGGCTAGTCTGTACTCAGCACACTGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57003</th>\n",
       "      <td>ENSMUSG00000067352</td>\n",
       "      <td>Gm14149</td>\n",
       "      <td>TCTTGCTTCTTGCTCTCTTGCTTCCTGCACCCTGGCTCCTGAAGATGTACGAAATAAAGCTTTGCCGCAGAAGATTCTGGTCTGTGGTGTTCTTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57004</th>\n",
       "      <td>ENSMUSG00000087337</td>\n",
       "      <td>Gm14144</td>\n",
       "      <td>TGTTCTTGGAGGACTCCTTTGGACCCCGAGGTTGTAAGCTCAAGCGGGTTGTTGTTTCAAATGAAAAGTAGGTAAATGTAGTAAATGTCTAAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57005</th>\n",
       "      <td>ENSMUSG00000089575</td>\n",
       "      <td>Gm23005</td>\n",
       "      <td>TCTACACCTAGAGGCATTGTTGCTTTGGCTCTGTGTTCCATGTGTATCAGTAGAAGGCAGAGAAGAGAGAGGCTAGTCTGTACTCAGCACTCTGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57006</th>\n",
       "      <td>ENSMUSG00000119806</td>\n",
       "      <td>Gm24077</td>\n",
       "      <td>CCTACACCTAGAGGCATTGTAGCTTTGGCTCTGTGTTCCATGTGTATCAGTAGAAGGCAGAGAAGAGAGAGGCTAGTCTGTATTCAGCACTCTGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57007</th>\n",
       "      <td>ENSMUSG00000087609</td>\n",
       "      <td>4930442J19Rik</td>\n",
       "      <td>TTGCTCCTGCGCAAAGGCCTCTGGTTGAAGCCGTTGGTGGGCAGTGCTGCTGCCTGTTCTTGGAGGACTCCTTTGGACCCCGAGGTTGTAAGCTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57008</th>\n",
       "      <td>ENSMUSG00000119525</td>\n",
       "      <td>Gm23261</td>\n",
       "      <td>CCTACCCCTAGAGGCATTGTAGCTTTGGCTCTGTGTTCCATGTGTATCAGTAGAAGGCAGAGAAGAGAGAGGCTAGTCTGTATTCAGCACTCTGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57009</th>\n",
       "      <td>ENSMUSG00000083391</td>\n",
       "      <td>Gm14148</td>\n",
       "      <td>ATGGTGAAGGTCGGTGTGAACGGATTTGGCCGAATTGGGCGCCTGGTCACCAGGGCTGCCATCTGCTCTCCACATGGCAAAGTGGAGATTGTTGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57010 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene stable ID      Gene name  \\\n",
       "0      ENSMUSG00000064336          mt-Tf   \n",
       "1      ENSMUSG00000064337        mt-Rnr1   \n",
       "2      ENSMUSG00000064338          mt-Tv   \n",
       "3      ENSMUSG00000064339        mt-Rnr2   \n",
       "4      ENSMUSG00000064340         mt-Tl1   \n",
       "5      ENSMUSG00000064341         mt-Nd1   \n",
       "6      ENSMUSG00000064342          mt-Ti   \n",
       "7      ENSMUSG00000064343          mt-Tq   \n",
       "8      ENSMUSG00000064344          mt-Tm   \n",
       "9      ENSMUSG00000064345         mt-Nd2   \n",
       "...                   ...            ...   \n",
       "57000  ENSMUSG00000079005        Gm14147   \n",
       "57001  ENSMUSG00000085805        Gm14145   \n",
       "57002  ENSMUSG00000119215        Gm22517   \n",
       "57003  ENSMUSG00000067352        Gm14149   \n",
       "57004  ENSMUSG00000087337        Gm14144   \n",
       "57005  ENSMUSG00000089575        Gm23005   \n",
       "57006  ENSMUSG00000119806        Gm24077   \n",
       "57007  ENSMUSG00000087609  4930442J19Rik   \n",
       "57008  ENSMUSG00000119525        Gm23261   \n",
       "57009  ENSMUSG00000083391        Gm14148   \n",
       "\n",
       "                                                                                                  Sequence  \n",
       "0                                     GTTAATGTAGCTTAATAACAAAGCAAAGCACTGAAAATGCTTAGATGGATAATTGTATCCCATAAACA  \n",
       "1      AAAGGTTTGGTCCTGGCCTTATAATTAATTAGAGGTAAAATTACACATGCAAACCTCCATAGACCGGTGTAAAATCCCTTAAACATTTACTTAAAA...  \n",
       "2                                    CATAGTGTAGCTTAATATTAAAGCATCTGGCCTACACCCAGAAGATTTCATGACCAATGAACACTCTGA  \n",
       "3      ACTAATCCTAGCCCTAGCCCTACACAAATATAATTATACTATTATATAAATCAAAACATTTATCCTACTAAAAGTATTGGAGAAAGAAATTCGTAC...  \n",
       "4                              ATTAGGGTGGCAGAGCCAGGAAATTGCGTAAGACTTAAAACCTTGTTCCCAGAGGTTCAAATCCTCTCCCTAATA  \n",
       "5      GTGTTCTTTATTAATATCCTAACACTCCTCGTCCCCATTCTAATCGCCATAGCCTTCCTAACATTAGTAGAACGCAAAATCTTAGGGTACATACAA...  \n",
       "6                                    AGAAATATGTCTGATAAAAGAATTACTTTGATAGAGTAAATTATAGAGGTTCAAGCCCTCTTATTTCTA  \n",
       "7                                  TAGGATAAGGTGTTTAGGTAGCACGGAGAATTTTGAATTCTTAAGTGTAGGTTCAATTCCTATTGTCCTAG  \n",
       "8                                    AGTAAGGTCAGCTAATTAAGCTATCGGGCCCATACCCCGAAAACGTTGGTTTAAATCCTTCCCGTACTA  \n",
       "9      ATAAATCCTATCACCCTTGCCATCATCTACTTCACAATCTTCTTAGGTCCTGTAATCACAATATCCAGCACCAACCTAATACTAATATGAGTAGGC...  \n",
       "...                                                                                                    ...  \n",
       "57000  GGAGAGTGGTTTGGTGGTGGTTTGGGAGACGTGGACTGTGGTGATAATTTTGGATTTGGATACTGGTAACGTTTTGGACTCAGATTCTGGCAAAGG...  \n",
       "57001  TGAGAACAGTTTCATCTGCATGGTGAACAGTGCAGGAGATGCCTGCCAAGCCCTAACACAGTAGACTGAAGATGGCCACACAGGTTATGCTTCCTT...  \n",
       "57002  CCCACACCTAGAGGCATTGTTGCTTTGGCTCTGTGTTCCATGTGTATCAGTAGAAGGCAGAGAAGAGAGAGGCTAGTCTGTACTCAGCACACTGAC...  \n",
       "57003  TCTTGCTTCTTGCTCTCTTGCTTCCTGCACCCTGGCTCCTGAAGATGTACGAAATAAAGCTTTGCCGCAGAAGATTCTGGTCTGTGGTGTTCTTCC...  \n",
       "57004  TGTTCTTGGAGGACTCCTTTGGACCCCGAGGTTGTAAGCTCAAGCGGGTTGTTGTTTCAAATGAAAAGTAGGTAAATGTAGTAAATGTCTAAAATT...  \n",
       "57005  TCTACACCTAGAGGCATTGTTGCTTTGGCTCTGTGTTCCATGTGTATCAGTAGAAGGCAGAGAAGAGAGAGGCTAGTCTGTACTCAGCACTCTGAC...  \n",
       "57006  CCTACACCTAGAGGCATTGTAGCTTTGGCTCTGTGTTCCATGTGTATCAGTAGAAGGCAGAGAAGAGAGAGGCTAGTCTGTATTCAGCACTCTGAC...  \n",
       "57007  TTGCTCCTGCGCAAAGGCCTCTGGTTGAAGCCGTTGGTGGGCAGTGCTGCTGCCTGTTCTTGGAGGACTCCTTTGGACCCCGAGGTTGTAAGCTCA...  \n",
       "57008  CCTACCCCTAGAGGCATTGTAGCTTTGGCTCTGTGTTCCATGTGTATCAGTAGAAGGCAGAGAAGAGAGAGGCTAGTCTGTATTCAGCACTCTGAC...  \n",
       "57009  ATGGTGAAGGTCGGTGTGAACGGATTTGGCCGAATTGGGCGCCTGGTCACCAGGGCTGCCATCTGCTCTCCACATGGCAAAGTGGAGATTGTTGCC...  \n",
       "\n",
       "[57010 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_df_unspliced = pd.read_pickle(UNSPLICED_RESULT_PATH)\n",
    "\n",
    "_df_unspliced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2960899"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_df_unspliced['Sequence'].str.len().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "for datum in SeqIO.parse(EXON_PATH, \"fasta\"):\n",
    "    parts = datum.id.split('|')\n",
    "    assert len(parts) == 2\n",
    "    parts.append(str(datum.seq))\n",
    "    records.append(parts)\n",
    "\n",
    "df = pd.DataFrame(records, columns=['Gene stable ID', 'Exon stable ID', 'Sequence'])\n",
    "\n",
    "df = _df_features.merge(df, how='outer')\n",
    "\n",
    "assert not df['Exon stable ID'].duplicated().any()\n",
    "\n",
    "df.to_pickle(EXON_RESULT_PATH)\n",
    "\n",
    "_df_exon = df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Exon stable ID</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000064336</td>\n",
       "      <td>mt-Tf</td>\n",
       "      <td>ENSMUSE00000521514</td>\n",
       "      <td>GTTAATGTAGCTTAATAACAAAGCAAAGCACTGAAAATGCTTAGATGGATAATTGTATCCCATAAACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000064337</td>\n",
       "      <td>mt-Rnr1</td>\n",
       "      <td>ENSMUSE00000521515</td>\n",
       "      <td>AAAGGTTTGGTCCTGGCCTTATAATTAATTAGAGGTAAAATTACACATGCAAACCTCCATAGACCGGTGTAAAATCCCTTAAACATTTACTTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000064338</td>\n",
       "      <td>mt-Tv</td>\n",
       "      <td>ENSMUSE00000521516</td>\n",
       "      <td>CATAGTGTAGCTTAATATTAAAGCATCTGGCCTACACCCAGAAGATTTCATGACCAATGAACACTCTGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000064339</td>\n",
       "      <td>mt-Rnr2</td>\n",
       "      <td>ENSMUSE00000521517</td>\n",
       "      <td>ACTAATCCTAGCCCTAGCCCTACACAAATATAATTATACTATTATATAAATCAAAACATTTATCCTACTAAAAGTATTGGAGAAAGAAATTCGTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000064340</td>\n",
       "      <td>mt-Tl1</td>\n",
       "      <td>ENSMUSE00000521518</td>\n",
       "      <td>ATTAGGGTGGCAGAGCCAGGAAATTGCGTAAGACTTAAAACCTTGTTCCCAGAGGTTCAAATCCTCTCCCTAATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000064341</td>\n",
       "      <td>mt-Nd1</td>\n",
       "      <td>ENSMUSE00000521519</td>\n",
       "      <td>GTGTTCTTTATTAATATCCTAACACTCCTCGTCCCCATTCTAATCGCCATAGCCTTCCTAACATTAGTAGAACGCAAAATCTTAGGGTACATACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000064342</td>\n",
       "      <td>mt-Ti</td>\n",
       "      <td>ENSMUSE00000521520</td>\n",
       "      <td>AGAAATATGTCTGATAAAAGAATTACTTTGATAGAGTAAATTATAGAGGTTCAAGCCCTCTTATTTCTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000064343</td>\n",
       "      <td>mt-Tq</td>\n",
       "      <td>ENSMUSE00000521521</td>\n",
       "      <td>TAGGATAAGGTGTTTAGGTAGCACGGAGAATTTTGAATTCTTAAGTGTAGGTTCAATTCCTATTGTCCTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000064344</td>\n",
       "      <td>mt-Tm</td>\n",
       "      <td>ENSMUSE00000521522</td>\n",
       "      <td>AGTAAGGTCAGCTAATTAAGCTATCGGGCCCATACCCCGAAAACGTTGGTTTAAATCCTTCCCGTACTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000064345</td>\n",
       "      <td>mt-Nd2</td>\n",
       "      <td>ENSMUSE00000521523</td>\n",
       "      <td>ATAAATCCTATCACCCTTGCCATCATCTACTTCACAATCTTCTTAGGTCCTGTAATCACAATATCCAGCACCAACCTAATACTAATATGAGTAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537972</th>\n",
       "      <td>ENSMUSG00000087609</td>\n",
       "      <td>4930442J19Rik</td>\n",
       "      <td>ENSMUSE00001092671</td>\n",
       "      <td>CTCTGGCCCTACTTCTCTGCCAATCAGCAGGAGCCTCTTCCACCAAGGAGATCTGAAAACAAGAGGTTTTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537973</th>\n",
       "      <td>ENSMUSG00000087609</td>\n",
       "      <td>4930442J19Rik</td>\n",
       "      <td>ENSMUSE00001017163</td>\n",
       "      <td>TGTTCTTGGAGGACTCCTTTGGACCCCGAGGTTGTAAGCTCAAGCGGGTTGTTGTTTCAAATGAAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537974</th>\n",
       "      <td>ENSMUSG00000087609</td>\n",
       "      <td>4930442J19Rik</td>\n",
       "      <td>ENSMUSE00000987498</td>\n",
       "      <td>ATGACAGTATTTCTTGAGACATCTTACTGGAGAAGTGGTGCCCATGCCTAAGAGACTGAGGGGGCAGTAACACCAATTGACTCTCCTTGCACCATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537975</th>\n",
       "      <td>ENSMUSG00000087609</td>\n",
       "      <td>4930442J19Rik</td>\n",
       "      <td>ENSMUSE00001057441</td>\n",
       "      <td>GTTCTTATTCATGATCATGTTTGGAACCAGATCCAGATATGAATTCAACTGTCAAACACTCCAGACTTCTCCTTGACTCCACTCAGAGGACAGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537976</th>\n",
       "      <td>ENSMUSG00000087609</td>\n",
       "      <td>4930442J19Rik</td>\n",
       "      <td>ENSMUSE00000822311</td>\n",
       "      <td>TTGCTCCTGCGCAAAGGCCTCTGGTTGAAGCCGTTGGTGGGCAGTGCTGCTGCCTGTTCTTGGAGGACTCCTTTGGACCCCGAGGTTGTAAGCTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537977</th>\n",
       "      <td>ENSMUSG00000087609</td>\n",
       "      <td>4930442J19Rik</td>\n",
       "      <td>ENSMUSE00001054029</td>\n",
       "      <td>GTTCCTGGGTATCACACACTTGGTAACACTTCTGCCTAAGAAATTAAACTCTCAGCGCTCTGTGGACATTCTCATTGGCTAATCTTCACTTGAACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537978</th>\n",
       "      <td>ENSMUSG00000087609</td>\n",
       "      <td>4930442J19Rik</td>\n",
       "      <td>ENSMUSE00000843206</td>\n",
       "      <td>GTTCCTGGGTATCACACACTTGGTAACACTTCTGCCTAAGAAATTAAACTCTCAGCGCTCTGTGGACATTCTCATTGGCTAATCTTCACTTGAACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537979</th>\n",
       "      <td>ENSMUSG00000119525</td>\n",
       "      <td>Gm23261</td>\n",
       "      <td>ENSMUSE00011467448</td>\n",
       "      <td>CCTACCCCTAGAGGCATTGTAGCTTTGGCTCTGTGTTCCATGTGTATCAGTAGAAGGCAGAGAAGAGAGAGGCTAGTCTGTATTCAGCACTCTGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537980</th>\n",
       "      <td>ENSMUSG00000083391</td>\n",
       "      <td>Gm14148</td>\n",
       "      <td>ENSMUSE00000721440</td>\n",
       "      <td>ATGGTGAAGGTCGGTGTGAACGGATTTGGCCGAATTGGGCGCCTGGTCACCAGGGCTGCCATCTGCTCTCCACATGGCAAAGTGGAGATTGTTGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537981</th>\n",
       "      <td>ENSMUSG00000083391</td>\n",
       "      <td>Gm14148</td>\n",
       "      <td>ENSMUSE00001278330</td>\n",
       "      <td>GAAGGGCTCATGACCACGGTCCTTGCCATCACTGCCACCCAGAAGACTGTGGATGACCCCCCTGGAAAGCTGTGGCGAGAGGGTCGTGGGACTGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>537982 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene stable ID      Gene name      Exon stable ID  \\\n",
       "0       ENSMUSG00000064336          mt-Tf  ENSMUSE00000521514   \n",
       "1       ENSMUSG00000064337        mt-Rnr1  ENSMUSE00000521515   \n",
       "2       ENSMUSG00000064338          mt-Tv  ENSMUSE00000521516   \n",
       "3       ENSMUSG00000064339        mt-Rnr2  ENSMUSE00000521517   \n",
       "4       ENSMUSG00000064340         mt-Tl1  ENSMUSE00000521518   \n",
       "5       ENSMUSG00000064341         mt-Nd1  ENSMUSE00000521519   \n",
       "6       ENSMUSG00000064342          mt-Ti  ENSMUSE00000521520   \n",
       "7       ENSMUSG00000064343          mt-Tq  ENSMUSE00000521521   \n",
       "8       ENSMUSG00000064344          mt-Tm  ENSMUSE00000521522   \n",
       "9       ENSMUSG00000064345         mt-Nd2  ENSMUSE00000521523   \n",
       "...                    ...            ...                 ...   \n",
       "537972  ENSMUSG00000087609  4930442J19Rik  ENSMUSE00001092671   \n",
       "537973  ENSMUSG00000087609  4930442J19Rik  ENSMUSE00001017163   \n",
       "537974  ENSMUSG00000087609  4930442J19Rik  ENSMUSE00000987498   \n",
       "537975  ENSMUSG00000087609  4930442J19Rik  ENSMUSE00001057441   \n",
       "537976  ENSMUSG00000087609  4930442J19Rik  ENSMUSE00000822311   \n",
       "537977  ENSMUSG00000087609  4930442J19Rik  ENSMUSE00001054029   \n",
       "537978  ENSMUSG00000087609  4930442J19Rik  ENSMUSE00000843206   \n",
       "537979  ENSMUSG00000119525        Gm23261  ENSMUSE00011467448   \n",
       "537980  ENSMUSG00000083391        Gm14148  ENSMUSE00000721440   \n",
       "537981  ENSMUSG00000083391        Gm14148  ENSMUSE00001278330   \n",
       "\n",
       "                                                                                                   Sequence  \n",
       "0                                      GTTAATGTAGCTTAATAACAAAGCAAAGCACTGAAAATGCTTAGATGGATAATTGTATCCCATAAACA  \n",
       "1       AAAGGTTTGGTCCTGGCCTTATAATTAATTAGAGGTAAAATTACACATGCAAACCTCCATAGACCGGTGTAAAATCCCTTAAACATTTACTTAAAA...  \n",
       "2                                     CATAGTGTAGCTTAATATTAAAGCATCTGGCCTACACCCAGAAGATTTCATGACCAATGAACACTCTGA  \n",
       "3       ACTAATCCTAGCCCTAGCCCTACACAAATATAATTATACTATTATATAAATCAAAACATTTATCCTACTAAAAGTATTGGAGAAAGAAATTCGTAC...  \n",
       "4                               ATTAGGGTGGCAGAGCCAGGAAATTGCGTAAGACTTAAAACCTTGTTCCCAGAGGTTCAAATCCTCTCCCTAATA  \n",
       "5       GTGTTCTTTATTAATATCCTAACACTCCTCGTCCCCATTCTAATCGCCATAGCCTTCCTAACATTAGTAGAACGCAAAATCTTAGGGTACATACAA...  \n",
       "6                                     AGAAATATGTCTGATAAAAGAATTACTTTGATAGAGTAAATTATAGAGGTTCAAGCCCTCTTATTTCTA  \n",
       "7                                   TAGGATAAGGTGTTTAGGTAGCACGGAGAATTTTGAATTCTTAAGTGTAGGTTCAATTCCTATTGTCCTAG  \n",
       "8                                     AGTAAGGTCAGCTAATTAAGCTATCGGGCCCATACCCCGAAAACGTTGGTTTAAATCCTTCCCGTACTA  \n",
       "9       ATAAATCCTATCACCCTTGCCATCATCTACTTCACAATCTTCTTAGGTCCTGTAATCACAATATCCAGCACCAACCTAATACTAATATGAGTAGGC...  \n",
       "...                                                                                                     ...  \n",
       "537972                              CTCTGGCCCTACTTCTCTGCCAATCAGCAGGAGCCTCTTCCACCAAGGAGATCTGAAAACAAGAGGTTTTG  \n",
       "537973                                  TGTTCTTGGAGGACTCCTTTGGACCCCGAGGTTGTAAGCTCAAGCGGGTTGTTGTTTCAAATGAAAA  \n",
       "537974  ATGACAGTATTTCTTGAGACATCTTACTGGAGAAGTGGTGCCCATGCCTAAGAGACTGAGGGGGCAGTAACACCAATTGACTCTCCTTGCACCATA...  \n",
       "537975  GTTCTTATTCATGATCATGTTTGGAACCAGATCCAGATATGAATTCAACTGTCAAACACTCCAGACTTCTCCTTGACTCCACTCAGAGGACAGGTG...  \n",
       "537976  TTGCTCCTGCGCAAAGGCCTCTGGTTGAAGCCGTTGGTGGGCAGTGCTGCTGCCTGTTCTTGGAGGACTCCTTTGGACCCCGAGGTTGTAAGCTCA...  \n",
       "537977  GTTCCTGGGTATCACACACTTGGTAACACTTCTGCCTAAGAAATTAAACTCTCAGCGCTCTGTGGACATTCTCATTGGCTAATCTTCACTTGAACA...  \n",
       "537978  GTTCCTGGGTATCACACACTTGGTAACACTTCTGCCTAAGAAATTAAACTCTCAGCGCTCTGTGGACATTCTCATTGGCTAATCTTCACTTGAACA...  \n",
       "537979  CCTACCCCTAGAGGCATTGTAGCTTTGGCTCTGTGTTCCATGTGTATCAGTAGAAGGCAGAGAAGAGAGAGGCTAGTCTGTATTCAGCACTCTGAC...  \n",
       "537980  ATGGTGAAGGTCGGTGTGAACGGATTTGGCCGAATTGGGCGCCTGGTCACCAGGGCTGCCATCTGCTCTCCACATGGCAAAGTGGAGATTGTTGCC...  \n",
       "537981  GAAGGGCTCATGACCACGGTCCTTGCCATCACTGCCACCCAGAAGACTGTGGATGACCCCCCTGGAAAGCTGTGGCGAGAGGGTCGTGGGACTGCC...  \n",
       "\n",
       "[537982 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_df_exon = pd.read_pickle(EXON_RESULT_PATH)\n",
    "\n",
    "_df_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(FEATURE_PATH, sep='\\t')\n",
    "\n",
    "df = df.groupby(['Gene Synonym']).filter(lambda x: x['Gene name'].dropna().drop_duplicates().shape[0] == 1)\n",
    "\n",
    "df = df[['Gene Synonym', 'Gene name']]\n",
    "\n",
    "df = df.drop_duplicates()\n",
    "\n",
    "df = df.set_index(['Gene Synonym'])\n",
    "\n",
    "ds = df['Gene name']\n",
    "\n",
    "gene_synonym_gene_name = ds.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "for data in SeqIO.parse(PROTEIN_FASTA_PATH, \"fasta\"):\n",
    "    record={}\n",
    "    for index, value in enumerate(re.findall(r'(?:(?<=^)[^|]+|(?<=\\|)[^| ]+|(?<=\\|)[^ ]+)|(?<=\\ ).+?(?= \\w\\w\\=|$)', data.description)):\n",
    "        value = value.strip()\n",
    "        if index == 0:\n",
    "            record['db'] = value\n",
    "        elif index == 1:\n",
    "            record['UniqueIdentifier'] = value\n",
    "        elif index == 2:\n",
    "            record['EntryName'] = value\n",
    "        elif index == 3:\n",
    "            record['ProteinName'] = value\n",
    "        else:\n",
    "            k, v = re.split(r'\\=', value)\n",
    "            record[k] = v\n",
    "\n",
    "    record['seq'] = str(data.seq)\n",
    "    \n",
    "    records.append(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(records)\n",
    "\n",
    "df['gene_symbol_harmonized'] = df['GN'].apply(lambda x: gene_synonym_gene_name.get(x, x))\n",
    "\n",
    "df = df.set_index('gene_symbol_harmonized').reset_index()\n",
    "\n",
    "_df_protein_sequence = df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = _df_protein_sequence.copy()\n",
    "\n",
    "df = df[['gene_symbol_harmonized', 'db', 'UniqueIdentifier', 'EntryName', 'ProteinName', 'OS', 'OX', 'GN', 'PE', 'SV', 'seq']]\n",
    "\n",
    "assert not df['UniqueIdentifier'].duplicated().any()\n",
    "\n",
    "df.to_pickle(PROTEIN_RESULT_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "development",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:24:37) \n[GCC 9.4.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a2d6514f9963693ddf9848f840dcbc46c3e7b9abcac09ab97b6beeb9a964f0bf"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
